snapshots/rawblock/rawblock.go

```diff
@@ -0,0 +1,548 @@
+
+	path := o.getSnapshotPath(id, kind)
+	renamed := o.getSnapshotPath("rm-"+id, kind)
+	if err := os.Rename(path, renamed); err != nil {
+		if !os.IsNotExist(err) {
+			return errors.Wrap(err, "failed to rename")
+		}
+		renamed = ""
+	}
+
+	err = t.Commit()
+	t = nil
+	if err != nil {
+		if renamed != "" {
+			if err1 := os.Rename(renamed, path); err1 != nil {
+				// May cause inconsistent data on disk
+				log.G(ctx).WithError(err1).WithField("path", renamed).Errorf("failed to rename after failed commit")
+			}
+		}
+		return errors.Wrap(err, "failed to commit")
+	}
+	if renamed != "" {
+		if err := os.RemoveAll(renamed); err != nil {
+			// Must be cleaned up, any "rm-*" could be removed if no active transactions
+			log.G(ctx).WithError(err).WithField("path", renamed).Warnf("failed to remove snapshot file")
+		}
+	}
+
+	return nil
+}
+
+// Walk all snapshots in the snapshotter. For each snapshot in the
+// snapshotter, the function will be called.
+func (o *snapshotter) Walk(ctx context.Context, fn func(context.Context, snapshots.Info) error) error {
+	ctx, t, err := o.ms.TransactionContext(ctx, false)
+	if err != nil {
+		return err
+	}
+	defer t.Rollback()
+	return storage.WalkInfo(ctx, fn)
+}
+
+func (o *snapshotter) createBaseImage(ctx context.Context) (baseImage string, err error) {
+	var base string
+	fd, err := ioutil.TempFile(filepath.Join(o.root, snapshots.KindActive.String()), "new-")
+	if err != nil {
+		return "", err
+	}
+	base = fd.Name()
+	fd.Close()
+	defer func() {
+		if err != nil {
+			if err1 := os.Remove(base); err1 != nil {
+				log.G(ctx).WithError(err1).Warnf("failed to remove temp file %s after base image creation failure", base)
+			}
+		}
+	}()
+	if err := os.Chmod(base, 0755); err != nil {
+		return "", err
+	}
+	if err := os.Truncate(base, int64(o.sizeMB<<20)); err != nil {
+		return "", err
+	}
+	switch o.fstype {
+	case "xfs":
+		if out, err := exec.Command("mkfs.xfs", "-f", base).CombinedOutput(); err != nil {
+			return "", errors.Errorf("Failed to create %s file system on %s: %v:%s", o.fstype, base, err, string(out))
+		}
+	case "ext4":
+		args := []string{
+			base,
+			"-F",
+			"-O",
+			"^uninit_bg",
+			"-E",
+			"nodiscard,lazy_itable_init=0,lazy_journal_init=0",
+		}
+		if out, err := exec.Command("mkfs.ext4", args...).CombinedOutput(); err != nil {
+			return "", errors.Errorf("Failed to create %s file system on %s: %v:%s", o.fstype, base, err, string(out))
+		}
+		// Remove ext4's lost+found as we want an empty directory instead
+		mount.WithTempMount(ctx, []mount.Mount{{Source: base, Type: o.fstype, Options: []string{"loop"}}},
+			func(root string) error {
+				return os.Remove(filepath.Join(root, "lost+found"))
+			})
+	default:
+		return "", errors.Errorf("unsupported fstype %s", o.fstype)
+	}
+
+	return base, err
+}
+
+func (o *snapshotter) copyReflinkImage(source, target string) error {
+	if out, err := exec.Command("cp", "-a", "--reflink=auto", source, target).CombinedOutput(); err != nil {
```

Comment:
can we use `continuity/fs.CopyFile`?
It supports reflink: https://github.com/containerd/continuity/blob/004b46473808b3e7a4a3049c20e4376c91eb966d/fs/copy_linux.go#L85

Reply:
Thanks for the pointer! I'll change to use it.